server/src/main/java/org/candlepin/model/ProductCurator.java

```diff
@@ -168,24 +158,38 @@ public Product getByName(Owner owner, String name) {
             return new HashSet<>();
         }
 
+        // Determine what is already in the L2 cache and load it directly. Multiload the remainder.
+        // This is because of https://hibernate.atlassian.net/browse/HHH-12944 where multiload ignores the
+        // L2 Cache.
         Set<Product> products = new HashSet<>();
-        Set<String> notInCache = new HashSet<>();
-        Cache<String, Product> productCache = candlepinCache.getProductCache();
-
-        //First find all products that are in cache. Those keys that
-        //are not present in the cache will not be in the result Map
-        Map<String, Product> productsFromCache = productCache.getAll(productUuids);
-        products.addAll(productsFromCache.values());
-
-        notInCache.addAll(productUuids);
-        notInCache.removeAll(productsFromCache.keySet());
-
-        //Now find, hydrate and cache all the products that has been
-        //missing in the cache
-        Map<String, Product> freshProducts = this.getHydratedProductsByUuid(notInCache);
+        Set<String> productsNotInCache = new HashSet<>();
+        for (String uuid :productUuids) {
+            if (currentSession().getSessionFactory().getCache().contains(this.entityType(), uuid)) {
+                products.add(currentSession().get(Product.class, uuid));
+            }
+            else {
+                productsNotInCache.add(uuid);
+            }
+        }
+        if (productsNotInCache.size() > 0) {
+            log.debug("Loading objects that were not already in the cache: " + productsNotInCache.size());
+            Session session = this.currentSession();
+            java.util.List entities = session.byMultipleIds(this.entityType())
+                .enableSessionCheck(true)
+                .multiLoad(new ArrayList(productsNotInCache));
```

Comment:
It bugs me that Hibernate's object here doesn't accept a generic collection, but at least they provide a variadic alternative. It's slightly faster here to use ```productsNotInCache.toArray()``` rather than building another collection.